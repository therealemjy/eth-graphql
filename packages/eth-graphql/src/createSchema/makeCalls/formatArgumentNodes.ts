import { ArgumentNode, GraphQLResolveInfo, Kind, ValueNode } from 'graphql';

import { SolidityValue } from '../../types';
import { ContractCallArgs } from './types';

// Extract a node's value
const getNodeValue = ({
  valueNode,
  variableValues,
}: {
  valueNode: ValueNode;
  variableValues: GraphQLResolveInfo['variableValues'];
}): SolidityValue => {
  if (valueNode.kind === Kind.NULL) {
    return null;
  }

  // Get variable node value
  if (valueNode.kind === Kind.VARIABLE) {
    const variableName = valueNode.name.value;
    return variableValues[variableName] as SolidityValue;
  }

  // Convert list to array
  if (valueNode.kind === Kind.LIST) {
    return valueNode.values.map(node =>
      getNodeValue({
        valueNode: node,
        variableValues,
      }),
    );
  }

  // Convert object to object
  if (valueNode.kind === Kind.OBJECT) {
    return valueNode.fields.reduce(
      (accObject, field) => ({
        ...accObject,
        [field.name.value]: getNodeValue({
          valueNode: field.value,
          variableValues,
        }),
      }),
      {},
    );
  }

  return valueNode.value;
};

const formatArgumentNodes = ({
  argumentNodes,
  variableValues,
}: {
  argumentNodes: ReadonlyArray<ArgumentNode>;
  variableValues: GraphQLResolveInfo['variableValues'];
}) =>
  argumentNodes.reduce<ContractCallArgs>(
    (accArguments, argument) => ({
      ...accArguments,
      [argument.name.value]: getNodeValue({
        valueNode: argument.value,
        variableValues,
      }),
    }),
    {},
  );

export default formatArgumentNodes;
