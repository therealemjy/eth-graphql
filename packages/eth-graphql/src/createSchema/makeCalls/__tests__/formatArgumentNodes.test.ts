import { ArgumentNode, Kind } from 'graphql';

import formatArgumentNodes from '../formatArgumentNodes';

describe('createSchema/makeCalls/formatArgumentNodes', () => {
  it('converts argument nodes to valid arguments', () => {
    const fakeVariableName = 'fakeVariable';
    const argumentNodes: ArgumentNode[] = [
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 0',
        },
        value: {
          kind: Kind.LIST,
          values: [
            {
              kind: Kind.LIST,
              values: [
                {
                  kind: Kind.STRING,
                  value: 'fake string 0',
                },
              ],
            },
          ],
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 1',
        },
        value: {
          kind: Kind.OBJECT,
          fields: [
            {
              kind: Kind.OBJECT_FIELD,
              name: {
                kind: Kind.NAME,
                value: 'fake name 2',
              },
              value: {
                kind: Kind.OBJECT,
                fields: [
                  {
                    kind: Kind.OBJECT_FIELD,
                    name: {
                      kind: Kind.NAME,
                      value: 'fake name 3',
                    },
                    value: {
                      kind: Kind.STRING,
                      value: 'fake string 1',
                    },
                  },
                ],
              },
            },
          ],
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 4',
        },
        value: {
          kind: Kind.STRING,
          value: 'fake string 2',
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 5',
        },
        value: {
          kind: Kind.BOOLEAN,
          value: true,
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 6',
        },
        value: {
          kind: Kind.INT,
          value: '100',
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 7',
        },
        value: {
          kind: Kind.FLOAT,
          value: '0.1',
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 7',
        },
        value: {
          kind: Kind.ENUM,
          value: '0',
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 8',
        },
        value: {
          kind: Kind.VARIABLE,
          name: {
            kind: Kind.NAME,
            value: fakeVariableName,
          },
        },
      },
    ];

    const res = formatArgumentNodes({
      argumentNodes,
      variableValues: {
        [fakeVariableName]: 'fake variable value',
      },
    });

    expect(res).toMatchSnapshot();
  });
});
