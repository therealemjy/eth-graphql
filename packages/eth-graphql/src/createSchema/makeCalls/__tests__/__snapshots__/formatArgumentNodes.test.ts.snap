// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSchema/makeCalls/formatArgumentNodes converts argument nodes to valid arguments 1`] = `
{
  "fake name 0": [
    [
      "fake string 0",
    ],
  ],
  "fake name 1": {
    "fake name 2": {
      "fake name 3": "fake string 1",
    },
  },
  "fake name 4": "fake string 2",
  "fake name 5": true,
  "fake name 6": "100",
  "fake name 7": "0",
  "fake name 8": "fake variable value",
}
`;
